src/indexBy.ts

```diff
@@ -52,6 +67,32 @@ const _indexBy =
     }, {});
   };
 
+function indexByStrict<K extends PropertyKey, T>(
+  array: ReadonlyArray<T>,
+  fn: (item: T) => K
+): Partial<Record<K, T>>;
+
+function indexByStrict<K extends PropertyKey, T>(
+  fn: (item: T) => K
+): (array: ReadonlyArray<T>) => Partial<Record<K, T>>;
+
+function indexByStrict() {
+  return purry(_indexByStrict(false), arguments);
+}
+
+const _indexByStrict =
+  (indexed: boolean) =>
+  <K extends PropertyKey, T>(
+    array: Array<T>,
+    fn: PredIndexedOptional<T, K>
+  ) => {
+    return array.reduce<Partial<Record<K, T>>>((ret, item, index) => {
+      const key = indexed ? fn(item, index, array) : fn(item);
+      ret[key] = item;
+      return ret;
+    }, {});
+  };
```

Comment:
We're not using the indexed version right now, so there's no point in making the code more complex.
```suggestion
  <K extends PropertyKey, T>(
    array: Array<T>,
    fn: (item: T) => K,
  ) => {
    return array.reduce<Partial<Record<K, T>>>((ret, item) => {
      const key = fn(item);
      ret[key] = item;
      return ret;
    }, {});
  };
```